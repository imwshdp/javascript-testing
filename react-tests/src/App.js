import { Outlet, Link } from 'react-router-dom';

function App() {
	return (
		<main className='app'>
			<div className='nav'>
				<Link data-testid='main-link' to={'/'}>
					Main Page
				</Link>
				<Link data-testid='about-link' to={'/about'}>
					About Page
				</Link>
				<Link data-testid='users-link' to={'/users'}>
					User list
				</Link>
			</div>

			<Outlet />
		</main>
	);
}

export default App;
