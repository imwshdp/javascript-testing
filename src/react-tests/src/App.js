import { useEffect, useState } from 'react';

function App() {
	const [data, setData] = useState(null);
	const [toggle, setToggle] = useState(false);
	const [inputValue, setInputValue] = useState('');

	const onClick = () => setToggle(previousToggle => !previousToggle);

	useEffect(() => {
		setTimeout(() => {
			setData({});
		}, 100);
	}, []);

	return (
		<div className='App'>
			<h1>Hello, world!</h1>

			<h2>Example with toggle</h2>
			{toggle && <div data-testid='toggle-element'>Toggled!</div>}
			<button data-testid='toggle-switcher' onClick={onClick}>
				Click me!
			</button>

			<h2>Example with input</h2>
			<input
				type='text'
				placeholder='input value'
				value={inputValue}
				onChange={event => setInputValue(event.target.value)}
			/>

			<h2>Example with async data fetching</h2>
			{data && <div style={{ color: 'red' }}>Data is received</div>}
		</div>
	);
}

export default App;
