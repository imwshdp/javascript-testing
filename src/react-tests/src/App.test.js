/* eslint-disable testing-library/no-debugging-utils */
import { fireEvent, render, screen } from '@testing-library/react';
import App from './App';

describe('Renders learn react link', () => {
	test('Examples with Get methods', () => {
		render(<App />);

		const helloWorldElement = screen.getByText(/Hello, world!/i);
		expect(helloWorldElement).toBeInTheDocument();

		const buttonElement = screen.getByRole('button');
		expect(buttonElement).toBeInTheDocument();

		const inputElement = screen.getByPlaceholderText('input value');
		expect(inputElement).toBeInTheDocument();
		expect(inputElement).toMatchSnapshot();
	});

	test('Examples with Query methods', () => {
		render(<App />);

		const nullableHelloWorldElement = screen.queryByText(/Hello, world! 2/i);
		expect(nullableHelloWorldElement).toBeNull();
	});

	test('Examples with Find methods', async () => {
		render(<App />);

		screen.debug();

		const dataElement = await screen.findByText(/Data is received/i);
		expect(dataElement).toBeInTheDocument();
		expect(dataElement).toHaveStyle({
			color: 'red'
		});

		screen.debug();
	});

	test('Examples with React click event', () => {
		render(<App />);

		const toggleSwitcher = screen.getByTestId('toggle-switcher');

		expect(screen.queryByTestId('toggle-element')).toBeNull();

		fireEvent.click(toggleSwitcher);
		expect(screen.getByTestId('toggle-element')).toBeInTheDocument();

		fireEvent.click(toggleSwitcher);
		expect(screen.queryByTestId('toggle-element')).toBeNull();
	});
});
