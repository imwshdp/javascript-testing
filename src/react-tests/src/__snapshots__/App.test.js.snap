// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders learn react link Basic JSX testing 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`Renders learn react link Examples with Get methods 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`renders learn react link 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;

exports[`renders learn react link renders learn react link 1`] = `
<input
  placeholder="input value"
  type="text"
/>
`;
